File: app/src/main/java/com/musicplayer/db/AppDatabase.java
package com.musicplayer.db;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import com.musicplayer.db.entity.Playlist;
import com.musicplayer.db.entity.PlaylistSong;
import com.musicplayer.db.entity.SearchHistory;

@Database(entities = {Playlist.class, PlaylistSong.class, SearchHistory.class}, version = 1)
public abstract class AppDatabase extends RoomDatabase {
    private static AppDatabase instance;
    public abstract MusicDao musicDao();

    public static synchronized AppDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(),
                    AppDatabase.class, "music_database")
                    .fallbackToDestructiveMigration()
                    .build();
        }
        return instance;
    }
}

File: app/src/main/java/com/musicplayer/db/AppDatabase.kt
package com.musicplayer.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.musicplayer.db.entity.*

@Database(
    entities = [Playlist::class, PlaylistSong::class, SearchHistory::class, FeedSong::class, PlaybackHistory::class],
    version = 10,
    exportSchema = false
)
abstract class AppDatabase : RoomDatabase() {
    abstract fun musicDao(): MusicDao

    companion object {
        @Volatile
        private var instance: AppDatabase? = null

        private val MIGRATION_5_6 = object : Migration(5, 6) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("CREATE TABLE IF NOT EXISTS `playback_history` (`songId` TEXT NOT NULL, `title` TEXT, `artist` TEXT, `url` TEXT, `duration` INTEGER NOT NULL, `thumbnailUrl` TEXT, `timestamp` INTEGER NOT NULL, PRIMARY KEY(`songId`))")
            }
        }

        private val MIGRATION_6_7 = object : Migration(6, 7) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("ALTER TABLE `personalized_feed` ADD COLUMN `isHeader` INTEGER NOT NULL DEFAULT 0")
            }
        }

        private val MIGRATION_7_8 = object : Migration(7, 8) {
            override fun migrate(database: SupportSQLiteDatabase) {
                // Reservado para futuras optimizaciones de índices
            }
        }

        @JvmStatic
        fun getInstance(context: Context): AppDatabase {
            return instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "music_database"
                )
                .fallbackToDestructiveMigration()
                .build().also { instance = it }
            }
        }
    }
}
